tensorflow/core/kernels/segment_reduction_ops_gpu.cu.h

```diff
@@ -1029,6 +1050,362 @@ struct SparseSegmentGradFunctor<GPUDevice, T, Index, SegmentId> {
+          static_cast<Tindices>(sorted_indices[i]);
+    }
+  }
+}
+
+template <typename Tindex, typename EdgeIndicatorIter,
+          typename Tindices_internal, typename Tindices>
+Status LaunchScatterUniqueIndicesKernel(
+    const GPUDevice& d, Tindex nouter,
+    EdgeIndicatorIter sorted_indices_edge_indicator,       // [nouter]
+    const Tindices_internal* __restrict__ sorted_indices,  // [nouter]
+    const Tindex* __restrict__ sorted_indices_ids,         // [nouter]
+    Tindices* __restrict__ sorted_unique_indices) {        // [num_unique]
+  GpuLaunchConfig config = GetGpuLaunchConfig(
+      nouter, d,
+      &ScatterUniqueIndicesKernel<Tindex, EdgeIndicatorIter, Tindices_internal,
+                                  Tindices>,
+      /*dynamic_shared_memory_size=*/0, /*block_size_limit=*/0);
+  return GpuLaunchKernel(
+      ScatterUniqueIndicesKernel<Tindex, EdgeIndicatorIter, Tindices_internal,
+                                 Tindices>,
+      config.block_count, config.thread_per_block, 0, d.stream(), nouter,
+      sorted_indices_edge_indicator, sorted_indices, sorted_indices_ids,
+      sorted_unique_indices);
+}
+
+template <typename T, typename Tindices, typename Tsegmentids>
+struct SparseSegmentGradV2Functor<GPUDevice, T, Tindices, Tsegmentids> {
+  void operator()(OpKernelContext* context,
+                  SparseSegmentReductionOperation operation,
+                  typename TTypes<T>::ConstMatrix input_flat,
+                  typename TTypes<Tindices>::ConstVec indices_vec,
+                  typename TTypes<Tsegmentids>::ConstVec segment_vec,
+                  const TensorShape& dense_output_shape,
+                  typename AsyncOpKernel::DoneCallback done) {
+    const GPUDevice& device = context->eigen_gpu_device();
+
+    const int64_t nsegments = input_flat.dimension(0);
+    const int64_t ninner64 = input_flat.dimension(1);
+    const int64_t nouter64 = indices_vec.dimension(0);
+    // Note: nouter and ninner are not expected to be huge, so we use int32 to
+    // save memory bandwidth.
+    using Tindex = int32;
+    OP_REQUIRES_ASYNC(
+        context, nouter64 <= std::numeric_limits<Tindex>::max(),
+        errors::InvalidArgument("Indices vector of length ", nouter64,
+                                " is too large to fit in int32."),
+        done);
+    const Tindex nouter = static_cast<Tindex>(nouter64);
+    OP_REQUIRES_ASYNC(
+        context, ninner64 <= std::numeric_limits<Tindex>::max(),
+        errors::InvalidArgument("Inner data dimension of size ", ninner64,
+                                " is too large to fit in int32."),
+        done);
+    const Tindex ninner = static_cast<Tindex>(ninner64);
+
+    // Cast indices to 32-bit to save memory bandwidth (the cost of the cast is
+    // worth it because the vector is used multiple times).
+    // Note that we can assume int32 is safe because the op's dense_output_dim0
+    // input is always int32.
+    using Tindices_internal = int32;
```

Comment:
If it's always fixed, then why have it as a template parameter at all?

Reply:
I think I was just leaving open the possibility that int64 output support might be added in future. I've added the word "currently" to the comment.